cdap-ui/graphql/graphql.js

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+const { ApolloServer } = require('apollo-server-express');
+const { importSchema } = require('graphql-import');
+const merge = require('lodash/merge');
+const log4js = require('log4js');
+const { applicationResolvers } = require('./resolvers/applicationResolvers');
+const { namespaceResolvers } = require('./resolvers/namespaceResolvers');
+const { metadataResolvers } = require('./resolvers/metadataResolvers');
+const { programRecordResolvers } = require('./resolvers/programRecordResolvers');
+const { programRecordTypeResolvers } = require('./resolvers/type/programRecordTypeResolver');
+const { scheduleResolvers } = require('./resolvers/scheduleResolvers');
+const { statusResolvers } = require('./resolvers/statusResolvers');
+
+const log = log4js.getLogger('graphql');
+const env = process.env.NODE_ENV || 'production';
+
+const resolvers = merge(applicationResolvers,
```

Comment:
We need to merge all of the resolvers into a single object to pass that object to the Apollo Server

Reply:
Which is still a single level object right? May be we just need `_.assign`?

Reply:
This is the object that I get, and need, when using `merge`:

```
{ Query: 
   { applications: [AsyncFunction: applications],
     application: [AsyncFunction: application],
     namespaces: [AsyncFunction: namespaces],
     status: [AsyncFunction: status] },
  ApplicationRecord: { applicationDetail: [AsyncFunction: applicationDetail] },
  ApplicationDetail: 
   { metadata: [AsyncFunction: metadata],
     programs: [AsyncFunction: programs] },
  ProgramRecord: { __resolveType: [AsyncFunction: __resolveType] },
  Workflow: 
   { runs: [AsyncFunction: runs],
     schedules: [AsyncFunction: schedules] },
  ScheduleDetail: { nextRuntimes: [AsyncFunction: nextRuntimes] } }
```

This is the object that I get with `assign`

```
{ Query: { status: [AsyncFunction: status] },
  ApplicationRecord: { applicationDetail: [AsyncFunction: applicationDetail] },
  ApplicationDetail: { programs: [AsyncFunction: programs] },
  ProgramRecord: { __resolveType: [AsyncFunction: __resolveType] },
  Workflow: 
   { runs: [AsyncFunction: runs],
     schedules: [AsyncFunction: schedules] },
  ScheduleDetail: { nextRuntimes: [AsyncFunction: nextRuntimes] } }
```

There is only one resolver per type (e.g., `Query` only has `status`, but it should have more.

I do not think we should use `_.assign`

Reply:
So this whole thing can be simplified,
**applicationResolvers.js**

```
{
    applications: async (parent, args, context) => {
      const namespace = args.namespace;
      const options = resolversCommon.getGETRequestOptions();
      options['url'] = urlHelper.constructUrl(cdapConfig, `/v3/namespaces/${namespace}/apps`);
      context.namespace = namespace;
      return await resolversCommon.requestPromiseWrapper(options);
    },
  }
```

in **graphql.js**

```
const { applicationResolvers } = require('./resolvers/applicationResolvers');
const { namespaceResolvers } = require('./resolvers/namespaceResolvers');

...

const server = new ApolloServer({
  typeDefs,
  {
    Query: {
        ...applicationResolvers,
        ...namespaceResolvers
    }
  },
  introspection: env === 'production' ? false : true,
  playground: env === 'production' ? false : true,
});
```

We are doing `_.merge` in a lot of places where it can be combined with just a spread operator.

Reply:
Similarly to `assign`, the syntax that you suggested does not include all resolvers. 

I believe the problem is that for each type (e.g., `Query`, `ApplicationRecord`) there are multiple resolvers, which `merge` merges them instead of replacing them.